pkg/transport/custom_transform_collection.go

```diff
@@ -0,0 +1,260 @@
+			ctc.invalidateCacheEntryLocked(ctx, true, ct.Name, bindingName, oldGroupResource, "CustomTransformSpec .Group or .Resource changed", "newGroupResource", groupResource)
+		} else {
+			klog.FromContext(ctx).Error(nil, "Impossible condition: ctNameToSpec has an entry but grToTransformData does not and no change in GroupResource", "customTransformName", ct.Name, "groupResource", groupResource, "bindingName", bindingName)
+		}
+	}
+	ctc.ctNameToSpec[ct.Name] = ct.Spec
+	return removes
+}
+
+func ctSpecGroupResource(spec v1alpha1.CustomTransformSpec) metav1.GroupResource {
+	return metav1.GroupResource{Group: spec.APIGroup, Resource: spec.Resource}
+}
+
+func (ctc *customTransformCollection) parseRemovesAndUpdateStatus(ctx context.Context, ct *v1alpha1.CustomTransform) (removes []jsonpath.Query) {
+	logger := klog.FromContext(ctx)
+	ctCopy := ct.DeepCopy()
+	ctCopy.Status = v1alpha1.CustomTransformStatus{ObservedGeneration: ct.Generation}
+	for idx, queryS := range ct.Spec.Remove {
+		query, err := jsonpath.ParseQuery(queryS)
+		if err != nil {
+			ctCopy.Status.Errors = append(ctCopy.Status.Errors, fmt.Sprintf("Error in spec.remove[%d]: %s", idx, err.Error()))
+		} else if len(query) == 0 {
+			ctCopy.Status.Errors = append(ctCopy.Status.Errors, fmt.Sprintf("Invalid spec.remove[%d]: it identifies the whole object", idx))
+		} else {
+			removes = append(removes, query)
+		}
+	}
+	ctEcho, err := ctc.client.UpdateStatus(ctx, ctCopy, metav1.UpdateOptions{FieldManager: ControllerName})
+	if err != nil {
+		logger.Error(err, "Failed to write status of CustomTransform", "name", ct.Name, "resourceVersion", ct.ResourceVersion, "status", ctCopy.Status)
+	} else {
+		logger.V(4).Info("Wrote status of CustomTransform", "name", ct.Name, "resourceVersion", ctEcho.ResourceVersion, "observedGeneration", ctCopy.Status.ObservedGeneration)
+	}
+	return
+}
+
+// invalidateCacheEntryLocked removes the cached entry for the given GroupResource.
+// Caller asserts that this is being done because of some change to the CustomTransform having the given name.
+// `shouldHave` asserts that ctc.ctNameToSpec has an entry for this name.
+// triggerBindingName, if not empty, is the name of the Binding being processed when this change was noticed.
+// reason and extraLogArgs go into the debug log statements.
+// Caller asserts that the ctc's mutex is locked.
+func (ctc *customTransformCollection) invalidateCacheEntryLocked(ctx context.Context, shouldHave bool, ctName, triggerBindingName string, oldGroupResource metav1.GroupResource, reason string, extraLogArgs ...any) {
+	logger := klog.FromContext(ctx)
+	oldGRTransformData, had := ctc.grToTransformData[oldGroupResource]
+	if !had {
+		if shouldHave {
+			logger.Error(nil, "Impossible condition: ctc.ctNameToSpec has an entry for a CustomTransform but ctc.grToTransformData has no entry for the GroupResource", "customTransformName", ctName, "groupResource", oldGroupResource)
```

Comment:
If I understand the intention, the current state of the code should never get here, but if someone updates the code and does not properly maintain the maps, this can happen.
**If this is the case, then why do we say it's impossible? It would be a misuse of maps/mutex - applies to the other instance of similar logging too.** 

--
Generally in both this and the other instance of an "impossible" error being logged, I think maintaining expected correctness as an invariant through controlling the mutability of the maps by design is better, such as through not exposing the maps/mutex to developers in this struct, but protecting them through an interface and providing functionality that does not break the invariance. Though is not a necessity or a blocker.

Reply:
Again, it is not just the maps, the consistency invariants involve the values in the maps too.

Not expose to "developers"? This is code that developers can modify; it is unavoidably exposed to developers. I suspect you mean to suggest moving `customTranformCollection` to its own new package. That could be done. I did not do it here because it seemed heavier weight than necessary to me. I agree that we can consider this in follow-on work.

Reply:
Technically everything is exposed to KS developers, yes.

What I meant is protecting the maps and direct mutation through moving their implementation somewhere else (can still be the same package), then abstracting the implementation through an interface.

Then the developer of custom transforms imp can use this new data structure as a black-box interface, leaving its content and actual implementation "protected".

This is the closest we can get in Go to implementing a class with private fields then using the class in the same package. As the user of the class (but still a developer in the package) - you practically can modify the class, but you leave it be and only use its exposed fields/methods because the point behind it is the abstraction of some logically related functionality.

Reply:
If we do not trust developers of the `pkg/transform` package then introducing a Go interface will not achieve isolation --- the goofy developer can still do stuff like `.(*customTransformCollectionImpl)`.

Reply:
I have introduced such an internal interface.